code/test-utils/src/main/java/com/adobe/marketing/tester/util/JSONObjectAsserts.kt

```diff
@@ -0,0 +1,940 @@
+                        pathTree = null,
+                        exactMatchMode = exactMatchMode,
+                        shouldAssert = shouldAssert) && finalResult
+                }
+                else {
+                    fail("""
+                    Actual side's index $index has already been taken by a wildcard match. Verify the test setup for correctness.
+
+                    Expected: ${expected.opt(index)}
+
+                    Actual (remaining unmatched elements): ${unmatchedActualElements.map { it.second }}
+
+                    Key path: ${keyPathAsString(keyPath)}
+                """.trimIndent())
+                    finalResult = false
+                    continue
+                }
+            }
+        }
+
+        return finalResult
+    }
+
+    /**
+     * Performs a flexible comparison between the given [expected] and [actual] JSON objects, optionally using exact match
+     * or value type match modes based on the provided [pathTree]. In case of a mismatch and if [shouldAssert] is true,
+     * an assertion error is thrown.
+     *
+     * This method is capable of handling nested JSON objects, and it provides the key path on assertion failures for
+     * easier debugging. The [pathTree] allows for more granular matching behavior.
+     *
+     * @param expected The expected JSON object to compare.
+     * @param actual The actual JSON object to compare.
+     * @param keyPath A list representing the key path to the current value being compared, especially
+     * useful for nested JSON objects.
+     * @param pathTree A map representing specific paths within the JSON structure that should be compared in certain ways.
+     * The path tree can include specific keys and their corresponding comparison modes or nested path trees.
+     * @param exactMatchMode If true, performs an exact match comparison; otherwise, uses value type matching.
+     * @param shouldAssert Indicates whether an assertion error should be thrown if [expected] and [actual] are not equal.
+     * Defaults to true.
+     *
+     * @return Returns true if [expected] and [actual] are equal based on the matching mode and the [pathTree], otherwise returns false.
+     *
+     * @throws AssertionError If [shouldAssert] is true and the [expected] and [actual] JSON objects are not equal.
+     */
+    private fun assertFlexibleEqual(
+        expected: JSONObject?,
+        actual: JSONObject?,
+        keyPath: MutableList<Any>,
+        pathTree: Map<String, Any>?,
+        exactMatchMode: Boolean,
+        shouldAssert: Boolean = true): Boolean {
+        if (expected == null) {
+            return true
+        }
+        if (actual == null) {
+            if (shouldAssert) {
+                fail("""
+                    Expected JSON is non-nil but Actual JSON is nil.
+
+                    Expected: ${expected.toString()}
+
+                    Actual: ${actual.toString()}
+
+                    Key path: ${keyPathAsString(keyPath)}
```

Comment:
can you remove the spacing between these lines, the logging looks off on failure